### PR TITLE
8294842: Pass actual pending cards to G1Policy::update_young_length_bounds during young gen revise

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that passes the actual amount of pending cards to `G1Policy::update_young_length_bounds` instead of some prediction during young gen revising?
That improves predictions quite a bit.

Depends on PR#10647, so please review that one first.

Testing: gha, local testing, tier1-5 with other changes in this patch series, perf testing with other changes in this patch series

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294842](https://bugs.openjdk.org/browse/JDK-8294842): Pass actual pending cards to G1Policy::update_young_length_bounds during young gen revise

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [ec5e3c7b](https://git.openjdk.org/jdk/pull/10649/files/ec5e3c7b47f9a1271d2f7f6ddc85a082b5f7eac4)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [ec5e3c7b](https://git.openjdk.org/jdk/pull/10649/files/ec5e3c7b47f9a1271d2f7f6ddc85a082b5f7eac4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10649/head:pull/10649` \
`$ git checkout pull/10649`

Update a local copy of the PR: \
`$ git checkout pull/10649` \
`$ git pull https://git.openjdk.org/jdk pull/10649/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10649`

View PR using the GUI difftool: \
`$ git pr show -t 10649`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10649.diff">https://git.openjdk.org/jdk/pull/10649.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1Policy.cpp [hunk 1]
```diff
@@ -194,11 +194,14 @@ uint G1Policy::calculate_desired_eden_length_by_mmu() const {
 }
 
 void G1Policy::update_young_length_bounds() {
-  update_young_length_bounds(_analytics->predict_rs_length());
+  // We have no measure of the number of pending cards in the thread buffers,
+  // assume these are very few.
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1Policy.cpp:198]: `predict_pending_cards()` includes the cards that were thread buffers.  It is based on the cards that were processed, which is done after DCQS::concatenate_logs.

File: src/hotspot/share/gc/g1/g1Policy.cpp [hunk 2]
```diff
@@ -523,7 +525,11 @@ void G1Policy::revise_young_list_target_length_if_necessary(size_t rs_length) {
     // add 10% to avoid having to recalculate often
     size_t rs_length_prediction = rs_length * 1100 / 1000;
     update_rs_length_prediction(rs_length_prediction);
-    update_young_length_bounds(rs_length_prediction);
+
+    G1DirtyCardQueueSet& dcqs = G1BarrierSet::dirty_card_queue_set();
+    // We have no measure of the number of cards in the thread buffers, assume
+    // these are very few compared to the ones in the DCQS.
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1Policy.cpp:531]: I think it's really that the #cards in thread buffers is very small compared to the sum of the other two sources.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1Policy.cpp (hunk 2 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1Policy.cpp:531]: Note that this comment does not say that what is assumed here is actually valid in all cases, but for now it suffices I guess.

File: src/hotspot/share/gc/g1/g1Policy.cpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1Policy.cpp:198]: Looking at the code:
`G1Policy::record_concurrent_refinement_stats` sets `_pending_cards_at_gc_start = dcqs.num_cards()`; `record_concurrent_refinement_stats` is called in `G1Policy::record_young_collection_start`; in `G1YoungCollector::collect`, `record_young_collection_start` is called _before_ `pre_evacuate_collection_set` which calls `G1BarrierSet::dirty_card_queue_set().concatenate_logs`.

I seem to be missing something, can you clarify?

Comment (kimbarrett) [reply src/hotspot/share/gc/g1/g1Policy.cpp:198]: That's a bug.  I even found and fixed it in some earlier version of my concurrent refinement changes.  But it looks like I never extracted it out or even created a bug report for it, and there's no fix for it in my concurrent refinement PR.

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1Policy.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/g1/g1Policy.cpp:198]: I'll file a bug for that problem.  I'd prefer this comment be removed, so it doesn't need to be tracked down for removal as part of fixing that bug.  Your call on that.

**Comment by tschatzl** [comment]
Thanks @kimbarrett @albertnetymk for your reviews

/integrate

**Pull request closed without merging**